### PR TITLE
WIP - Add the functionality to store head information in Constituent.

### PR DESCRIPTION
**Pull request by bhargav**
Tested with respect to ACEReader changes.

Will add documentation and run some more tests.

FYI @danyaljj

### CODE REVIEW

File: core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/datastructures/textannotation/Constituent.java [hunk 1]
```diff
@@ -88,8 +94,22 @@ public Constituent(String label, String viewName, TextAnnotation text, int start
...
+        this(label, 1.0, viewName, text, start, end, headStart, headEnd, true);
+    }
+
+    public Constituent(String label, double score, String viewName, TextAnnotation text, int start,
+                       int end, int headStart, int headEnd) {
+        this(label, score, viewName, text, start, end, headStart, headEnd, true);
+    }
+
+    private Constituent(String label, double score, String viewName, TextAnnotation text, int start,
+                       int end, int headStart, int headEnd, boolean hasHeadInformation) {
```
Comment (danyaljj) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/datastructures/textannotation/Constituent.java]: Why not make `hasHeadInformation`, `private` and with default `false` value. Whenever you use the constructers which accept head information you make `hasHeadInformation`, `true`.

File: core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/datastructures/textannotation/Constituent.java [hunk 2]
```diff
@@ -145,6 +163,49 @@ public Constituent(String label, double score, String viewName, TextAnnotation t
...
                 + ", " + endCharOffset + ")";
+
+        this.hasHeadInformation = hasHeadInformation;
+
+        IntPair headSpan = new IntPair(headStart, headEnd);
+        this.headSpan = headSpan;
+        int headStartSpan = headSpan.getFirst();
+        int headEndSpan = headSpan.getSecond();
+
+        if (headStart >= 0) {
```
Comment (danyaljj) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/datastructures/textannotation/Constituent.java]: `headStart >= 0 ||  headEnd >= 0` ?

File: core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/datastructures/textannotation/Constituent.java (hunk 1 above)
Comment (bhargav) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/datastructures/textannotation/Constituent.java]: I can't make it `final` then. I can make that change if required.

**Comment by bhargav** [comment]
Also, I can't seem to create the `"test.ta"` file. I get an exception in curated related call.

**Comment by danyaljj** [comment]
huh ...  what?

**Comment by danyaljj** [comment]
Oh I see you need to deserialize it again since the TextAnnotation has changed...

**Comment by bhargav** [comment]
Obsolete change now. Not required.

**Pull request closed without merging**